app/src/org/commcare/gis/MapboxLocationPickerActivity.kt

```diff
@@ -0,0 +1,206 @@
+package org.commcare.gis
+
+import android.content.Intent
+import android.os.Bundle
+import android.view.View
+import androidx.core.content.ContextCompat
+import androidx.lifecycle.ViewModelProvider
+import com.mapbox.android.core.location.LocationEngineCallback
+import com.mapbox.android.core.location.LocationEngineResult
+import com.mapbox.mapboxsdk.camera.CameraPosition
+import com.mapbox.mapboxsdk.camera.CameraUpdateFactory
+import com.mapbox.mapboxsdk.geometry.LatLng
+import com.mapbox.mapboxsdk.location.LocationComponentActivationOptions
+import com.mapbox.mapboxsdk.location.modes.CameraMode
+import com.mapbox.mapboxsdk.location.modes.RenderMode
+import com.mapbox.mapboxsdk.maps.MapboxMap
+import com.mapbox.mapboxsdk.maps.Style
+import com.mapbox.mapboxsdk.plugins.annotation.Symbol
+import com.mapbox.mapboxsdk.plugins.annotation.SymbolManager
+import com.mapbox.mapboxsdk.plugins.annotation.SymbolOptions
+import kotlinx.android.synthetic.main.activity_mapbox_location_picker.*
+import org.commcare.activities.components.FormEntryConstants
+import org.commcare.dalvik.R
+import org.commcare.utils.GeoUtils
+import org.commcare.views.widgets.GeoPointWidget
+
+class MapboxLocationPickerActivity : BaseMapboxActivity() {
+
+    companion object {
+        const val MARKER_ICON_IMAGE_ID = "marker_icon_image"
+    }
+
+    private lateinit var viewModel: MapboxLocationPickerViewModel
+    private var loadedStyle: Style? = null
+    private val mapStyles = arrayOf(
+            Style.MAPBOX_STREETS,
+            Style.SATELLITE_STREETS
+    )
+    private var currentMapStyleIndex = 0
+    private var cameraPosition: CameraPosition? = null
+    private var symbolManager: SymbolManager? = null
+    private var symbol: Symbol? = null
+    private val mapClickListener = MapboxMap.OnMapClickListener { point ->
+        // Add marker.
+        updateMarker(point)
+        viewModel.reverseGeocode(point)
+        true
+    }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        viewModel = ViewModelProvider(this, ViewModelProvider.AndroidViewModelFactory.getInstance(application))
+                .get(MapboxLocationPickerViewModel::class.java)
+        attachListener()
+
+        // Check extras
+        if (intent.hasExtra(GeoPointWidget.LOCATION)) {
+            val inputLocation = intent.getDoubleArrayExtra(GeoPointWidget.LOCATION)!!
+            cameraPosition = CameraPosition.Builder()
+                    .target(LatLng(inputLocation[0], inputLocation[1]))
+                    .zoom(10.0)
+                    .tilt(20.0)
+                    .build()
+        }
+        mapView.showCurrentLocationBtn(false)
```

Comment:
I wanted to get a callback on the button click and didn't find any docs on how to do that. So I disabled the button and added our own button on top of map layer to handle the callbacks.

Reply:
I am confused why do you need a listener on this button. Should not you utilize the location change listener instead ?

Reply:
Did you mean using `locationProvider` to listen to location change events? That'd probably add a lot of code and using a custom button was much easier.

Reply:
I see, I thought we would already be using location listeners to update the user location. Do user need to click the my location button everytime he wants to update the location ? That seems contradictory to our approach with the normal google maps based feature set where we seem to be attaching a [location listener](https://github.com/dimagi/commcare-android/blob/master/app/src/org/commcare/activities/GeoPointMapActivity.java#L164). Thoughts ?

Reply:
What do you mean when you say `update the location`? I feel that the map will always start from user's location and then user can move the marker anywhere, but after that when you wanna move the marker back to their current position, than they will use that `my location button` which will simply move the marker and nothing else.

Reply:
We should be consistent between both map activity in regards to when and how we update the location of the user. From reading code it seems like we update the location as soon as we have a more accurate location than the last time in the normal map activity in addition to the user click. Is that correct inference from me ? If yes, we should have the new map activity also incorporate that.

Reply:
Ohh okay, so I should keep listening to location from the provider. 
About the button thing, I'd still need the custom button since kujaku mapview doesn't provide the listener for button click.

Reply:
I looked at the normal map activity and there is no listener set on the my location button, instead the marker should be getting set by the location listener itself. Is there a reason we still need the my location listener after we add a listener ?

Reply:
Here's the listener in the normal map activity https://github.com/dimagi/commcare-android/blob/master/app/src/org/commcare/activities/GeoPointMapActivity.java#L309

The reason why we need a listener for button click : 

We'll keep updating the current location from the listener so long as user doesn't interact with the map. Once a user moves the map and puts the marker somewhere, then we no longer update the marker position from the listener. We'll then need a listener for the my location button so that after changing the marker, if user decides to move the marker back to the current position using that button, we should be able to do so.

Reply:
yes that reasoning makes sense to me.